src/main/java/org/swisspush/reststorage/FilePutter.java

```diff
@@ -0,0 +1,194 @@
+    private int moveToFinalDestinationAttemptCount = 0;
+
+    /**
+     * Package-private because currently only used internally.
+     */
+    FilePutter(Vertx vertx, String fullPath, FilePutterCallbacks callbacks, Handler<Resource> onCompleteHandler) {
+        this.vertx = vertx;
+        this.fullPath = canonicalizeAbsolutePath(fullPath);
+        this.callbacks = callbacks;
+        this.onCompleteHandler = onCompleteHandler;
+    }
+
+    /**
+     * <p>Triggers the configured task. This method should only get called once per
+     * instance!</p>
+     *
+     * @throws IllegalStateException Eg. in case method gets called more than one time.
+     */
+    public synchronized void execute() {
+        if (executed) {
+            throw new IllegalStateException("This putter already got executed.");
+        }
+        final FileSystem fileSystem = vertx.fileSystem();
+        this.executed = true;
+        // Setup required context.
+        this.tmpFilePath = "/.tmp/uploads/" + new File(fullPath).getName() + "-" + UUID.randomUUID().toString() + ".part";
+        this.tmpFilePathAbs = callbacks.canonicalize(tmpFilePath);
+        this.tmpFileParentPath = new File(tmpFilePathAbs).getParent();
+        // Prepare directory for temporary file.
+        fileSystem.mkdirs(tmpFileParentPath, result -> {
+            if (result.succeeded()) {
+                openTmpFile();
+            } else {
+                log.warn("Failed to create directory '" + tmpFileParentPath + "'.");
+                resolveWithErroneousResource();
+            }
+        });
+    }
+
+    private void openTmpFile() {
+        final FileSystem fileSystem = vertx.fileSystem();
+        fileSystem.open(tmpFilePathAbs, new OpenOptions(), result -> {
+            if (result.succeeded()) {
+                resolveWithTmpFileResource(result.result());
+            } else {
+                log.warn("Failed to open tmp file '{}'.", tmpFilePathAbs);
+                resolveWithErroneousResource();
+            }
+        });
+    }
+
+    private void resolveWithTmpFileResource(final AsyncFile tmpFile) {
+        final DocumentResource d = new DocumentResource();
+        d.writeStream = tmpFile;
+        d.closeHandler = v -> {
+            tmpFile.close(ev -> {
+                moveTmpFileToFinalDestination(d);
+            });
+        };
+        d.addErrorHandler(err -> {
+            log.error("Put file failed:", err);
+            cleanupFile(tmpFile);
+        });
+        // Resolve with ready-to-use resource.
+        onCompleteHandler.handle(d);
+    }
+
+    private void moveTmpFileToFinalDestination(DocumentResource d) {
+        final FileSystem fileSystem = vertx.fileSystem();
+        if (moveTimerHandle == 0) {
+            // Start a timer observing if we time out with our retries.
+            moveTimerHandle = vertx.setTimer(MOVE_RETRY_TIMEOUT_MILLIS, aLong -> {
+                moveOperationTimedOut = true;
```

Comment:
Interesting approach with the timed flag.
But why not simply store System.currentTimeMillis() in an instance variable?

Reply:
Simply because I wrote it this way without searching for alternative ways to do the same thing :). We could of course change that. Does this matter somehow?